pkgs/development/compilers/dmd/default.nix

```
@@ -1,199 +1,163 @@
-      homepage = http://dlang.org/;
-      # Everything is now Boost licensed, even the backend.
-      # https://github.com/dlang/dmd/pull/6680
-      license = licenses.boost;
-      maintainers = with maintainers; [ ThomasMader ];
-      platforms = [ "x86_64-linux" "i686-linux" "x86_64-darwin" ];
-    };
-  };
+  # https://issues.dlang.org/show_bug.cgi?id=19553
+  hardeningDisable = [ "fortify" ];
+
+  postUnpack = ''
+      patchShebangs .
+  '';
+
+  postPatch = stdenv.lib.optionalString stdenv.hostPlatform.isLinux ''
+      substituteInPlace phobos/std/socket.d --replace "assert(ih.addrList[0] == 0x7F_00_00_01);" ""
+  '';
+
+  nativeBuildInputs = [ ldc makeWrapper unzip which gdb git ]
+
+  ++ stdenv.lib.optional stdenv.hostPlatform.isDarwin (with darwin.apple_sdk.frameworks; [
+    Foundation
+  ]);
 
-  # Need to test Phobos in a fixed-output derivation, otherwise the
-  # network stuff in Phobos would fail if sandbox mode is enabled.
-  #
-  # Disable tests on Darwin for now because of
-  # https://github.com/NixOS/nixpkgs/issues/41099
-  phobosUnittests = if !stdenv.hostPlatform.isDarwin then
-    stdenv.mkDerivation rec {
-      name = "phobosUnittests-${version}";
-      version = dmdBuild.version;
-
-      enableParallelBuilding = dmdBuild.enableParallelBuilding;
-      preferLocalBuild = true;
-      inputString = dmdBuild.outPath;
-      outputHashAlgo = "sha256";
-      outputHash = builtins.hashString "sha256" inputString;
-
-      srcs = dmdBuild.srcs;
-
-      sourceRoot = ".";
-
-      nativeBuildInputs = dmdBuild.nativeBuildInputs;
-      buildInputs = dmdBuild.buildInputs;
-
-      buildPhase = ''
-          cd phobos
-          echo ${tzdata}/share/zoneinfo/ > TZDatabaseDirFile
-          echo ${curl.out}/lib/libcurl.so > LibcurlPathFile
-          make -j$NIX_BUILD_CORES -f posix.mak unittest BUILD=release ENABLE_RELEASE=1 PIC=1 DMD=${dmdBuild}/bin/dmd DFLAGS="-version=TZDatabaseDir -version=LibcurlPath -J$(pwd)"
-      '';
-
-      installPhase = ''
-          echo -n $inputString > $out
-      '';
-    }
+  buildInputs = [ curl tzdata ];
+
+  bits = builtins.toString stdenv.hostPlatform.parsed.cpu.bits;
+  osname = if stdenv.hostPlatform.isDarwin then
+    "osx"
+  else
+    stdenv.hostPlatform.parsed.kernel.name;
+  top = "$(echo $NIX_BUILD_TOP)";
+  pathToDmd = "${top}/dmd/generated/${osname}/release/${bits}/dmd";
+
+  curlExtension = if stdenv.hostPlatform.isDarwin then
+    ".4.dylib"
+  else
+    ".so";
+
+  # Buid and install are based on http://wiki.dlang.org/Building_DMD
+  buildPhase = ''
+      cd dmd
+      make -j$NIX_BUILD_CORES -f posix.mak INSTALL_DIR=$out BUILD=release ENABLE_RELEASE=1 PIC=1 HOST_DMD=ldmd2
+      cd ../druntime
+      make -j$NIX_BUILD_CORES -f posix.mak BUILD=release ENABLE_RELEASE=1 PIC=1 INSTALL_DIR=$out DMD=${pathToDmd}
+      cd ../phobos
+      echo ${tzdata}/share/zoneinfo/ > TZDatabaseDirFile
+      echo ${curl.out}/lib/libcurl${curlExtension} > LibcurlPathFile
```

```suggestion
      echo ${curl.out}/lib/libcurl${stdenv.hostPlatform.extensions.sharedLibrary} > LibcurlPathFile
```
Please excuse my ignorance, but don't know why you need ".4" before "dylib". If you really need it, you could use readlink:
```suggestion
      echo `readlink "${curl.out}/lib/libcurl${stdenv.hostPlatform.extensions.sharedLibrary}"` > LibcurlPathFile
```